lightning/src/routing/scoring.rs

```diff
@@ -460,25 +453,466 @@ impl<T: Time> Readable for ChannelFailure<T> {
+		*self.min_liquidity_offset_msat = if amount_msat < self.min_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.min_liquidity_offset_msat)
+		};
+		*self.last_updated = self.now;
+	}
+}
+
+const FRACTIONAL_BITS: u32 = 3;
+const FRACTIONAL_BITMASK: u64 = (1 << FRACTIONAL_BITS) - 1;
+const LOG2_FRACTIONAL_PART: [f64; 1 << FRACTIONAL_BITS] =
+	[0.0, 0.17, 0.32, 0.46, 0.58, 0.70, 0.81, 0.91];
+const LOG2_10: f64 = 3.32;
+
+fn log10_approx(numerator: u64, denominator: u64) -> f64 {
+	(log2_approx(numerator) - log2_approx(denominator)) / LOG2_10
+}
+
+#[inline]
+fn log2_approx(x: u64) -> f64 {
+	let leading_zeros = x.leading_zeros();
+	let integer_part = (63 - leading_zeros) as f64;
+	let fractional_part = LOG2_FRACTIONAL_PART[
+		(((x << leading_zeros) >> (63 - FRACTIONAL_BITS)) & FRACTIONAL_BITMASK) as usize
+	];
+	integer_part + fractional_part
+}
+
+fn log10_approx_without_fractional(numerator: u64, denominator: u64) -> f64 {
+	(log2_approx_without_fractional(numerator) - log2_approx_without_fractional(denominator)) / LOG2_10
+}
+
+#[inline]
+fn log2_approx_without_fractional(x: u64) -> f64 {
+	(63 - x.leading_zeros()) as f64
+}
+
+impl<G: Deref<Target = NetworkGraph>, T: Time> Score for ProbabilisticScorerUsingTime<G, T> {
+	fn channel_penalty_msat(
+		&self, short_channel_id: u64, amount_msat: u64, capacity_msat: u64, source: &NodeId,
+		target: &NodeId
+	) -> u64 {
+		if *source == self.node_id || *target == self.node_id {
+			return 0;
+		}
+
+		let liquidity_penalty_multiplier_msat = self.params.liquidity_penalty_multiplier_msat;
+		let liquidity_offset_half_life = self.params.liquidity_offset_half_life;
+		let success_probability = self.channel_liquidities
+			.get(&short_channel_id)
+			.unwrap_or(&ChannelLiquidity::new())
+			.as_directed(source, target, capacity_msat, liquidity_offset_half_life)
+			.success_probability(amount_msat);
+		match success_probability {
+			Probability::Zero => u64::max_value(),
+			Probability::One => 0,
+			Probability::Ratio { numerator, denominator } => {
+				let log_success_probability = log10_approx(numerator, denominator);
+				(-log_success_probability * liquidity_penalty_multiplier_msat as f64) as u64
+			},
+		}
+	}
+
+	fn payment_path_failed(&mut self, path: &[&RouteHop], short_channel_id: u64) {
+		let amount_msat = path.split_last().map(|(hop, _)| hop.fee_msat).unwrap_or(0);
+		let liquidity_offset_half_life = self.params.liquidity_offset_half_life;
+		let network_graph = self.network_graph.read_only();
+		let hop_sources = core::iter::once(self.node_id)
+			.chain(path.iter().map(|hop| NodeId::from_pubkey(&hop.pubkey)));
+		for (source, hop) in hop_sources.zip(path.iter()) {
+			let target = NodeId::from_pubkey(&hop.pubkey);
+			if source == self.node_id || target == self.node_id {
+				continue;
+			}
+
+			let capacity_msat = network_graph.channels()
+				.get(&hop.short_channel_id)
+				.and_then(|channel| channel.as_directed_to(&target).map(|d| d.effective_capacity()))
```

Comment:
Yep, I assume this is the same point made in the following conversation?

https://github.com/lightningdevkit/rust-lightning/pull/1227/files#r787934743

Reply:
Yes it is the same - only that here is the point where it is actually being initialized. I think this should really be fixed. While stuffing several small HTLCs into a channel in the context of MPP is one concern I think the much bigger concern is that the probability computation becomes severely wrong. Especially large channels should be highly attractive / likely. But I guess folks with large channels will want to cap the max size of htlcs which means that you would potentially not find such channels that have a high chance for high liquidity.

Reply:
This may be worth fixing, though in practice today its a relatively rare issue. Its also worth noting that not all users *have* chain data and don't bother fetching the channel's capacity by looking at the chain, using `htlc_maximum_msat` instead to determine it, because its the best info they have. Some of our users aren't the only ones doing this, I believe Electrum does this as well. Thus, we *really* need the implementation here to be robust even if we're using `htlc_maximum_msat` data to infer channel sizes, I'm not sure how best to do that, but just ensuring we pipe the values from on-chain through doesn't really solve the concern.

Ultimately this needs to be resolved at the spec level, of course, but given the values are basically always the same today we may just wait until we have a new gossip format for that.

Reply:
Addressed this as per a top-level comment, but dropped it in a favor of a follow-up as per offline discussion. Commit was a bit more involved and had a performance regressions, though pre-computing the `DirectedChannelInfo::htlc_maximum_msat` seem to fix it, oddly given it should only be calculated once.

Reply:
FWIW dropping all of the branches (even hitting the log10 call for 0) is very marginally faster in the benchmarks than either having both branches or only the `==0` branch.